### PR TITLE
STM32H7: manage Dual core boot whatever option bytes configuration

### PR DESCRIPTION
**Pull request by ABOSTM**
STM32H7: manage Dual core boot whatever option bytes configuration

Manage Dual core boot automatically whatever Option Bytes configuration.
No more need of KConfig STM32H7_DUAL_CORE_BOOT to match Option Bytes.

Signed-off-by: Alexandre Bourdiol <alexandre.bourdiol@st.com>

### CODE REVIEW

File: soc/arm/st_stm32/stm32h7/soc_m7.c
Comment (erwango) [inline soc/arm/st_stm32/stm32h7/soc_m7.c:29]:
```suggestion
		/* CM4 is started at boot in parallel of CM7
```

**Review by erwango — CHANGES_REQUESTED** [review]
Nit pick, otherwise LTGM

File: soc/arm/st_stm32/stm32h7/soc_m7.c
Comment (ABOSTM) [reply soc/arm/st_stm32/stm32h7/soc_m7.c]: Done

**Review by erwango — APPROVED** [review]

**Comment by erwango** [comment]
@ABOSTM can you rebase  ?

**Comment by erwango** [comment]
@Nukersson , @lochej can you have a look ?

**Comment by ABOSTM** [comment]
> @ABOSTM can you rebase ?

Done

**Comment by lochej** [comment]
> 
> 
> @Nukersson , @lochej can you have a look ?

LGTM, I'm not able to test this one as I don't have any STM32H745/757. But that implementation complies with the RM0399 when it comes to CM4 booting.

**Comment by erwango** [comment]
> > @Nukersson , @lochej can you have a look ?
> 
> LGTM, I'm not able to test this one as I don't have any STM32H745/757. But that implementation complies with the RM0399 when it comes to CM4 booting.

@lochej are you able to approve then ? We need 2 approvals to get a release engineer to merge it.

**Comment by Nukersson** [comment]
> @Nukersson , @lochej can you have a look ?

Tested the branch of  this PR on my NUCLEO-H745ZI-Q as follows:

1. Cortex-M7: running [Hello World](https://docs.zephyrproject.org/latest/samples/hello_world/README.html) over `USART3`
2. Cortex-M4: running [Blinky](https://docs.zephyrproject.org/latest/samples/basic/blinky/README.html#overview)

After the compilation both cores booting properly, `minicom` console shows `hello_world` and `LD2` blinks.

Moreover, setting two breakpoints in `soc_m7.c` on lines `47` and `54` respectively leads to break on line `47`, which means  following:

```
/* CM4 is started at boot in parallel of CM7
* but CM4 should set itself into stop mode,
* waiting for CM7 clock initialization.
*/
```

But I have only one problem (maybe incorrect understanding) with the implementation:

- What if developer does not need M4 booting anyway? As I understand the `else` of `if (READ_BIT(SYSCFG->UR1, SYSCFG_UR1_BCM4)) {` (line 54 of `soc_m7.c`) will force Cortex-M4 booting.

**Comment by ABOSTM** [comment]
> * What if developer does not need M4 booting anyway? As I understand the `else` of `if (READ_BIT(SYSCFG->UR1, SYSCFG_UR1_BCM4)) {` (line 54 of `soc_m7.c`) will force Cortex-M4 booting.

In such case developer should disable CONFIG_STM32H7_DUAL_CORE, and M4 wakeup will not happen any more:
```
#if defined(CONFIG_STM32H7_DUAL_CORE)
/* Unlock M4 once system configuration has been done */
SYS_INIT(stm32h7_m4_wakeup, POST_KERNEL, CONFIG_APPLICATION_INIT_PRIORITY);
#endif /* CONFIG_STM32H7_DUAL_CORE */
```

**Review by lochej — APPROVED** [review]

**Comment by Nukersson** [comment]
> > * What if developer does not need M4 booting anyway? As I understand the `else` of `if (READ_BIT(SYSCFG->UR1, SYSCFG_UR1_BCM4)) {` (line 54 of `soc_m7.c`) will force Cortex-M4 booting.
> 
> In such case developer should disable CONFIG_STM32H7_DUAL_CORE, and M4 wakeup will not happen any more:
> 
> ```
> #if defined(CONFIG_STM32H7_DUAL_CORE)
> /* Unlock M4 once system configuration has been done */
> SYS_INIT(stm32h7_m4_wakeup, POST_KERNEL, CONFIG_APPLICATION_INIT_PRIORITY);
> #endif /* CONFIG_STM32H7_DUAL_CORE */
> ```

Perfect form Alexander Kozhinov ( still learning Zephyr's KConfig 😄 )

**Review by KozhinovAlexander — APPROVED** [review]
LGTM

**Comment by lochej** [comment]
> > > @Nukersson , @lochej can you have a look ?
> > 
> > 
> > LGTM, I'm not able to test this one as I don't have any STM32H745/757. But that implementation complies with the RM0399 when it comes to CM4 booting.
> 
> @lochej are you able to approve then ? We need 2 approvals to get a release engineer to merge it.

I approve these changes but I don't have write access, so my approval doesn't count 😅

**Comment by erwango** [comment]
> I approve these changes but I don't have write access, so my approval doesn't count sweat_smile

@lochej Only a limited set of people (said "release engineers") have write access and are able to merge PRs once approved.
Though, anyone with interest in a change can review and provide feedback (approve, comment or request modification). Among those, some will be able to block a PR because they are codeowner, have an recognized expertise or generally known to provide valuable comments. Minimum of 2 approvals is required so a release engineer can merge a PR. So, the more the merrier...

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**
